### ISSUE TITLE
git link to self hosted platform

### ISSUE THREAD

**Issue by Fatsie** [hdl/awesome#172]
I am looking on adding [Coriolis](https://gitlab.lip6.fr/vlsi-eda/coriolis) to the list (PR will likely only be done after some doc changes have been done to Coriolis).  
French bureaucracy does not allow the Sorbonne university to host their (open source) projects on external sites like github or gitlab. That's the reason the LIP6 department is running their own gitlab instance. When I understand template.md correctly currently only github, gitlab and bitbucket are supported as source repository links.
I would like to have to possibility of having a general git link with possibly still a separate web link, something like:

    links:
      git: https://gitlab.lip6.fr/vlsi-eda/coriolis.git
      web: <...HOME_PAGE.html>

As I find the current code repository hosting too centralized, I am also considering moving my [Chips4Makers](https://www.chips4makers.io/) projects currently [hosted on gitlab](https://gitlab.com/Chips4Makers) to a self-hosted [Gitea](https://gitea.io/) instance. And in the mid-term I do want to have some of the projects in a state that makes them candidates for adding them to this awesome list...

**Comment by umarcor** [comment]
@Fatsie I understand your concerns and I agree with your suggestions.

The point about `gh` and `gl` is that they use different icons. That's the difference compared to `web`. In the case of self-hosted forges, what icon should we use?

- Use GitLab's, because those are self-hosted GitLab.
- Use a generic icon. E.g. https://materialdesignicons.com/icon/git.
- Support providing custom icons.

This is the part of the template/layout where this is defined: https://github.com/hdl/awesome/blob/46b1235c143f379b6b69d24962dd4abae05adf4f/layouts/partials/icons.html. The icons available in the theme are https://github.com/Templarian/MaterialDesign/tree/648ea94e9fce7faa40ca5b695cd2823d77f34743.

I propose using the generic git icon for all `git` entries in `links`.

BTW, providing multiple `gh`, `git` or `web` items in the `links` does not produce an error, but the last one is displayed only (see e.g. https://raw.githubusercontent.com/hdl/awesome/develop/content/items/superprove.md). I think we should support providing a list. For instance:

```
links:
  git: https://gitlab.lip6.fr/vlsi-eda/coriolis
  gh: coriolismirrororg/coriolis
  web:
    - URL_TO_MAIN_PAGE
    - URL_TO_ANOTHER_RELEVANT_PAGE
```

/cc @rodrigomelo9 

PS: @Fatsie if you are going to deploy your own Gitea, you might find Drone to be a nice self-hosted CI complement. See https://docs.drone.io/server/provider/gitea/.

**Comment by umarcor** [comment]
> And in the mid-term I do want to have some of the projects in a state that makes them candidates for adding them to this awesome list...

I overlooked this. I do honestly believe that your work is awesome already. If you don't feel confident for adding individual repos/projects to this list yet, please go ahead and submit a PR for "Chips4Makers" as a whole. In the description, you can provide references to stable and work-in-progress resources. Having https://hdl.github.io/awesome/items/chips4makers created will allow us to use it as an entrypoint to your whole ecosystem (talks, software experiments, contributions, articles...).

**Comment by rodrigomelo9** [comment]
Mmm. I don't know about to add `git` like a new type of link. We will have always a new problem (svn? sourceforge? a new one?). GitHub (mainly) and GitLab are pretty details because are commonly used.

I totally agree with multilink support (on `web`, but also `gh`, `gl` and `doc`?). We can check if single or multi and act in consequence, to avoid to modify the previous ones.

@umarcor let me know what do you think and I can perform the changes ;-)

Regards

**Comment by rodrigomelo9** [comment]
> I don't know about to add git like a new type of link. We will have always a new problem (svn? sourceforge? a new one?).

In the same matter, I think that we must remove `bb` (BitBucket) from `template.md`, which is actually not currently supported :lol:

**Comment by umarcor** [comment]
> Mmm. I don't know about to add `git` like a new type of link. We will have always a new problem (svn? sourceforge? a new one?). GitHub (mainly) and GitLab are pretty details because are commonly used.

> In the same matter, I think that we must remove bb (BitBucket) from template.md, which is actually not currently supported :lol:

Yes. The point is that we (materialdesign) have icons for GitHub and GitLab, but not for others (bb, or self-hosted). There is a generic icon for `git`, which we can use for bb and self-hosted git. That would not work for svn or mercurial. If we don't have specific icons for them, we can use some other (e.g. key `scm` and icon `source-repository`).

I think we should `bb` (and maybe `sf`) so we can write the project name only, and not the full URL. That's different from `git` and/or `scm`, where the full URL is required.

> I totally agree with multilink support (on `web`, but also `gh`, `gl` and `doc`?). We can check if single or multi and act in consequence, to avoid to modify the previous ones.

I think we can use the same logic (helper function/shortcode) for handling them. The "check if single or list" is independent from the icon or the type of string (project name or URL).

**Comment by umarcor** [comment]
This should have been closed automatically when #174 was merged. Anyway, links to git repositories in self-hosted platforms are supported now. See https://raw.githubusercontent.com/hdl/awesome/develop/template.md
